core/src/main/java/com/graphhopper/GraphHopper.java

```diff
@@ -861,7 +872,15 @@ public RoutingAlgorithmFactory getAlgorithmFactory(String profile, boolean disab
         if (chPreparationHandler.isEnabled() && !disableCH) {
             return chPreparationHandler.getAlgorithmFactory(profile);
         } else if (lmPreparationHandler.isEnabled() && !disableLM) {
-            return lmPreparationHandler.getAlgorithmFactory(profile);
+            for (LMProfileConfig lmp : lmPreparationHandler.getLMProfileConfigs()) {
+                if (lmp.getProfile().equals(profile)) {
+                    return lmp.usesOtherPreparation()
+                            // cross-querying
+                            ? lmPreparationHandler.getAlgorithmFactory(lmp.getPreparationProfile())
+                            : lmPreparationHandler.getAlgorithmFactory(lmp.getProfile());
```

Comment:
What about always using `getPreparationProfile` here that uses the method `usesOtherPreparation` internally?

Reply:
I thought about this also for a moment, but then I did not want to put too much logic into the getter, as it is also used by Jackson for deserialization, but I will have a second look.

Reply:
Ah, sure, I did not thought about this. Then we should indeed avoid this or rename the method (`resolveProfile`?)

Reply:
Hm, its only used once this way so I would keep it as is. `resolveProfile` is too similar to what we call 'resolving the profile' in `ProfileResolver` imo.

Reply:
Ah, ok. Sure.